library/alloc/src/vec/source_iter_marker.rs

```diff
@@ -1,3 +1,108 @@
+//! Inplace iterate-and-collect specialization for `Vec`
+//!
+//! This applies to iterators in the shape of
+//! `source.adapter().adapter().adapter().collect::<Vec<U>>()`
+//! where `source` is an owning iterator obtained from `Vec`, `Box<[]>` (by conversion to `Vec`)
+//! or `BinaryHeap`, the adapters each consume one or more items per step
+//! (represented by `InPlaceIterable`), provide transitive access to `source` (via `SourceIter`)
+//! and thus the underlying allocation. And finally the layouts of `source::Item` and `U` must
+//! have the same size and alignment, this is currently ensured via const eval instead of trait
+//! bounds.
+//!
+//! By extension some other collections which use `collect::Vec<_>()` internally in their
+//! `FromIterator` implementation benefit from this too.
+//!
+//! Access to the underlying source goes through a further layer of indirection via the private
+//! trait `AsIntoIter` to hide the implementation detail that other collections may use
+//! `vec::IntoIter` internally.
+//!
+//! In-place iteration performs the following steps
```

Comment:
You proceed to give some semi-complete description of implementation details how the specialized `from_iter` implementation in this module operates. The description seems hard (or impossible) to follow without also looking at the actual source code. AFAIK, explaining in words exactly what some source code does is often not a very useful form of documentation.

You intermix some short mentioning of relevant safety requirements, related issues, important points. Thus this is probably more than just the “description of implementation details” that I called it in the previous paragraph. I suppose, the heading “In-place iteration performs the following steps” doesn’t hint at the presence of all this extra information either, though. Some (or all?) of the additional information provided here seems like it might be better placed alongside the implementation itself.

**Overall, I’m wondering who is intended to read this text and for what reason.**

I can think of someone who
* wants to read and understand the specialization infrastructure and concrete implementation(s) of `.collect()`
  * _but_ they’d read the source code, too, so for them there’s probably lots of redundancy, and a lack of linking the explainations to the actual source code in these docs
* wants to get a feeling of which cases of calling `.collect()` results in what kind of runtime behavior regarding optimization of code and number of allocations
  * _but_ they wouldn’t care about the exact implementation and wouldn’t necessarily read some `alloc`-internal (private) module level docs in the first place
* wants to modify the implementations in or around this module
  * _but_ they’d prefer to work with concrete invariants attatched to concrete `unsafe` functions and traits
  * I guess they do appreciate your explicit mention of the fact that the current situation requires lots of “holistic”, “non-local” reasoning because it tells them that they won’t get all the relevant details regarding memory safety of the existing specializations in the form of concrete invariants on concrete functions/traits

<hr>

Let me show in an example how there’s a problem of unclear target audience for these docs as well as some general lack of details (or alternatively references to other places that fill in the details)

Take for example the phrase “For this to be sound the source has to use pointers too to perform its reads, otherwise there would be aliasing conflicts.” in step 1. From reading this sentence, I don’t know the exact location in the source code that this part discusses, I don’t know if there’s sufficient documentation on these soundness requirements on the trait infrastructure for `SourceIter` (and/or other traits) already and if yes where to find those. I’m not told that this isn’t discussing some generic properties of a widely-used trait infrastructure; instead “the source” is referring to just two types, `IntoIter` for `Vec` and for `BinaryHeap`.

And what to take from this “A has to do B” statement!? Does it tell me that it’s important to know that these two types use raw pointers directly? Great thing to point out in the right place in the source code, although I’m not actually sure what the best place would be. Possibly somewhere on/in the `SpecFromIter` impl in this very module. Or does it say that it’s important that we never change the fact that these two types use raw pointers? Maybe add a documentation note in the implementation of those `IntoIter` types. Provided there isn’t such a note already. Otherwise someone working on these `IntoIter` iterators may very well never learn about this requirement in the first place. This seems like a good instance of the “non-local” reasoning you mentioned and explaining it here doesn’t seem to add too much value. Maybe as a FIXME to find the right place to document this interaction: We do eventually want to get rid of any need for non-local reasoning here, right?

I could probably to the same kind of analysis like on this example to other parts of the text below, too, but I’m out of time for now, any it would probably be too much to discuss at once anyways.

Reply:
> Thus this is probably more than just the “description of implementation details” that I called it in the previous paragraph.

I could rewording the "performs the following steps" part then. I think the context of what is does helps to provide context for all the things that need to be considered.

Or do you think it would be better to just drop the information that this module requires the interaction of multiple unsafe traits to work correctly and then just leave the reader to figure out the details by reading the implementation and the documentation on those other traits?

> Overall, I’m wondering who is intended to read this text and for what reason.

It's not aimed at one specific group. Basically anyone involved in the standard library touching this area. This can be a reviewer, someone who explicitly wants to work on this or someone who needs to understand it in some other context such as the issues we had recently about unsoundness and needs a bird's eye view of what's happening here.

> and a lack of linking the explainations to the actual source code in these docs

The issue here is that afaik rustdoc has no mechanism to link to the source and some of the traits are `doc(hidden)` so they can't be linked to either. But I can give the latter a try anyway.

> Some (or all?) of the additional information provided here seems like it might be better placed alongside the implementation itself.

Except that "the implementation" is not a single place given all the moving parts that are involved. So to give an overview it has to be placed in one piece so that the reader doesn't have to collect all the individual pieces first to get a picture.

> I’m not told that this isn’t discussing some generic properties of a widely-used trait infrastructure; instead “the source” is referring to just two types, IntoIter for Vec and for BinaryHeap.

Well, I am considering to extend this to `VecDeque` since that should be feasible with the `make_contiguous()` function it gained a while ago. But that's speculative at the moment.

> We do eventually want to get rid of any need for non-local reasoning here, right?

We can certainly reduce it, but ultimately you'll at least have to reason about the interaction of multiple unsafe traits, a dozen implementations thereof and the safe things those implementations can do (such as the clones and panics mentioned). I think having a place that explains how those things fit together still helps even if in principle each piece already does its part in isolation.

> Or does it say that it’s important that we never change the fact that these two types use raw pointers? Maybe add a documentation note in the implementation of those IntoIter types. Provided there isn’t such a note already.

There isn't. Maybe `AsIntoIter` should be made unsafe, document that requirement and expose the methods needed for in-place iteration, that'll also help implementing this for `VecDeque`.


----

Anyway, I fail to take away many actionable items from this. I'll try adding some links and put some extra docs on `AsIntoIter`

Reply:
> The issue here is that afaik rustdoc has no mechanism to link to the source and some of the traits are `doc(hidden)` so they can't be linked to either. But I can give the latter a try anyway.

Regarding `doc(hidden)`, the linking may still work when you’re using something like
```
RUSTDOCFLAGS='--document-private-items --document-hidden-items' x.py doc library/std
```
to build the docs. Except if `doc(hidden)` items generate error messages when you intra-doc-link to them (from private items, which are usually not rendered either), I haven’t tested that much yet. Unfortunately I myself seem to have recently broken this command (with the `--document-hidden-items`) by replacing the import of `TrustedRandomAccess` with `TrustedRandomAccessNoCoerce` in a module that had a hidden doc link to `TrustedRandomAccess`… oops… maybe we should fix this and somehow add a test for this?

> Except that "the implementation" is not a single place given all the moving parts that are involved. So to give an overview it has to be placed in one piece so that the reader doesn't have to collect all the individual pieces first to get a picture.

Maybe this can be fixed… I just remembered something I’ve seen in the Haskell standard library. Take a look [here for example](https://hackage.haskell.org/package/containers-0.6.5.1/docs/src/Data.Map.Internal.html). There’s some named “[Note: …]” sections explaining implementation details that are lengthy and/or relevant at multiple places in the source code. Then, comments in the code refer to them by `See note: …`. Assuming that you’re unfamiliar with the lexical syntax of Haskell, note that these notes and the references to them are only part of comments, not any rendered documentation, and I personally don’t know if these “See notes: …” links have any sort of tooling that allows you to follow them when reading the source code or if they are just human convention. Anyways, my takeaway: We could copy this style/approach somewhat. Instead of the rather unstructured approach of just randomly addressing important soundness considerations hidden in the middle of a section describing itself as “In-place iteration performs the following steps”, you could give every notable aspect, a small headline \[e.g. **“Use of pointers for inplace-from_iter”**, etc…] and a self-contained description, and then simply refer to it from multiple relevant parts of the source code that all form “the implementatation”.



> Or do you think it would be better to just drop the information that this module requires the interaction of multiple unsafe traits to work correctly and then just leave the reader to figure out the details by reading the implementation and the documentation on those other traits?

AFAICT, the implementation is not all that long, so realistically reading the implementation is an approach that everyone trying to gaining an understanding of the implementation will attempt. Maybe all we need is to improve the inline docs of the implementation to be easier to understand without too much prior knowledge on all the traits involved, i.e. the approach could be: Make the implementation itself an even better and more readable alternative to the list you created here.

Even if we don’t do this, I think there could be: References in the implementation referring back to the “steps” laid out in the module-level docs (in order to link those more clearly, and also to keep this list of steps up to date); separating the description of what happens from discussion of safety considerations, either by completely separating the step-by-step description of what’s done from the following discussion, or at least by visibly separating the “what happens” part from the rest _within_ each of the steps listed. Also I wouldn’t mind if you added – on every aspect you’re discussing – an explanation whether that aspect is _fully_ covered/explained by documented safety-invariants on unsafe functions/traits, or if – instead – there’s non-local reasoning involved. The latter (i.e. explanations of some instance of non-local reasoning) should probably best be called out explicitly everywhere, so the reader knows to pay attention and maybe even to encourage them to think about future improvements that could make requirements more locally documented.



> We can certainly reduce it, but ultimately you'll at least have to reason about the interaction of multiple unsafe traits, a dozen implementations thereof and the safe things those implementations can do (such as the clones and panics mentioned). I think having a place that explains how those things fit together still helps even if in principle each piece already does its part in isolation.

I think you’re addressing two different things here. There’s definitely merit in explaining interactions of multiple unsafe traits, etc, even when each trait already has its own documentation with a correct complete list of safety requirements, etc. And then there’s explanations of implementation details that are only sound because we applied knowledge of lots of other implementation details of other parts in the standard library, i.e. the “non-local reasoning” we’re talking about. Making clear to the reader which is which seems valuable.

Regarding fixing the non-local stuff in the future, I think some reasonably low-effort approaches for improving the situation could be to (not mutually exclusive):
* Point out important details as you did, but put them in a place that can be referenced from multiple places, then reference them from all relevant places. (See the Haskell comparison above.)
* Make clear indications about which other parts in the standard library depend on implementation details of a function / struct / etc… especially across file boundaries. Similarly indicate when unsafe code relies on implementation details of other functions / structs / etc, especially if those other functions / types might seem unrelated. (And make sure to also add “back-references” in those implementations then.)
* For **private** (and often small) specialization traits or helper traits like `AsIntoIter`, consider documenting the trait with the exhaustive list of implementors and the exhaustive of users. (For `AsIntoIter` that’s 2 implementors and one use-case, I think.) This way we don’t necessarily need to come up with a complete list of safety invariants immediately, it could just say “This trait is implemented by `vec::IntoIter` and `binary_heap::IntoIter` and used exclusively for the `SpecFromIter` implementation in `vec/source_iter_marker.rs` where unsafe code depends on the implementation details of those types.” (Still, also marking the trait and its method `unsafe` would be useful, too.) Someone who wants to add implementations would either need to understand the use-case enough in order to soundly add an implementor (updating the documentation accordingly) or they would need to flesh out some explicit safety invariants first. Arriving at an intermediate state where _some_ invariants are listed (without claiming to be complete) and the use-cases and implementors are still listed is also reasonable.

Reply:
> Anyway, I fail to take away many actionable items from this.

Of course! I’m doing more of a writing-down-my-thoughts thing here, sorry in case that isn’t too helpful. I’ll probably look back into this tomorrow and see if I can come up with any more concrete suggestions for small, actionable improvements. I certainly don’t want you to _remove anything_ that you’ve written and considered relevant. More documentation is most often better than less, and it can also always be further improved or moved to a different place in the future.


> I'll try adding some links and put some extra docs on `AsIntoIter`

That sounds great! More “links” (and be it just in the form of “explaining where to look for more information”, not necessarily literal “links”) is probably useful, in particular to make things easier to understand in-depth. And I mentioned the suggestion to be more explicit on what things are just outlining things that are documented more in-depth elsewhere vs. what things are pointing out otherwise undocumented or barely documented non-local-reasoning kind of stuff.

Reply:
> Unfortunately I myself seem to have recently broken this command

In the meantime, `RUSTDOCFLAGS='--document-private-items --document-hidden-items -Arustdoc::broken-intra-doc-links' x.py doc library/std` helps.

Reply:
Thanks, I think I can work with some of that.

Regarding the notes, headlines should work fine for that since they're already linkable, as long as the module itself is linkable.